src/dataframe/dataframe.jl

```diff
@@ -846,16 +851,222 @@ function _deleteat!_helper(df::DataFrame, drop)
     return df
 end
 
+"""
+    keepat!(df::DataFrame, inds)
+
+Delete rows at all indices not specified by `inds` from a `DataFrame` `df`
+in place and return it.
+
+Internally `deleteat!` is called for all columns so `inds` must be:
+a vector of sorted and unique integers, a boolean vector, an integer,
+or `Not` wrapping any valid selector.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     2      5
+   3 │     3      6
+
+julia> keepat!(df, [1, 3])
+2×2 DataFrame
+ Row │ a      b
+     │ Int64  Int64
+─────┼──────────────
+   1 │     1      4
+   2 │     3      6
+```
+"""
+function keepat!(df::DataFrame, inds)
+    if !issorted(inds, lt=<=)
+        throw(ArgumentError("Indices passed to keepat! must be unique and sorted"))
+    end
```

Comment:
I thought I had left a comment but I can't find it again. This checks that indices are sorted twice. How about avoiding it by calling directly `_deleteat!_helper`, and doing the check here and in `deleteat!` instead?

Reply:
OK - I have fixed it and several other bugs, which lead me to noticing that:
```
julia> [1, 2, 3][Not([1.5])]
2-element Vector{Int64}:
 1
 2
```
(I reported the reason it works in #internals on Slack - let us see what response we get there; the code I prepared works around this issue for most common cases)